src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py

```diff
@@ -343,18 +343,39 @@ async def _handle_data_stream(
             stream_senders = self._get_metric_senders(
                 category, self._req_streaming_metrics[comp_id]
             )
-        api_data_receiver = self.comp_data_receivers[comp_id]
+        api_data_receiver: Receiver[Any] = self.comp_data_receivers[comp_id]
+
+        senders_done: asyncio.Event = asyncio.Event()
+        pending_messages = 0
 
         def process_msg(data: Any) -> None:
             tasks = []
             for extractor, senders in stream_senders:
                 for sender in senders:
                     tasks.append(sender.send(Sample(data.timestamp, extractor(data))))
-            asyncio.gather(*tasks)
+
+            def done_callback(_: asyncio.Future) -> None:
+                nonlocal pending_messages
+                pending_messages -= 1
+                if pending_messages == 0:
+                    senders_done.set()
+
+            nonlocal pending_messages
+            pending_messages += 1
+            asyncio.gather(*tasks).add_done_callback(done_callback)
 
         async for data in api_data_receiver:
+            senders_done.clear()
             process_msg(data)
 
+        while pending_messages > 0:
+            await senders_done.wait()
+
+        for _, requests in self._req_streaming_metrics[comp_id].items():
+            for request in requests:
+                # pylint: disable=protected-access
+                await self._registry._close_channel(request.get_channel_name())
```

Comment:
I don't like how obscure this is getting. 3 nested functions, `nonlocal`s and adding callbacks to a implicitly created task that is left to run wild seems like too much, at rleast for not getting some extra documentation.

You can remove some complexity making at least the done callback sequential (I didn't test it so it might have some bug, but *I think* it should work):

```py
        async def process_msg(data: Any) -> None:
            tasks = []
            for extractor, senders in stream_senders:
                for sender in senders:
                    tasks.append(sender.send(Sample(data.timestamp, extractor(data))))
            await asyncio.gather(*tasks)
            nonlocal pending_messages
            pending_messages -= 1
            if pending_messages == 0:
                senders_done.set()

        async for data in api_data_receiver:
            senders_done.clear()
            pending_messages += 1
            asyncio.create_task(process_msg(data))  # document this is left dangling on purpose

        while pending_messages > 0:
            await senders_done.wait()
```

Ideally we should keep track of running tasks and then wait for all at this point, instead of having a condition and the "global" counter, but I can understand it is some extra work so I would be fine with merging without that.

Reply:
Your code does indeed look nicer. I struggled with this a bit and the intention was to keep working on this bit in future PRs.
I also def. want to try a variant of this where we don't create a new task for every single receiver & sample.